activesupport/lib/active_support/claims.rb

```diff
@@ -0,0 +1,69 @@
+require 'active_support/time'
+
+module ActiveSupport
+  class Claims # :nodoc:
+    class InvalidClaims < StandardError; end
+    class ExpiredClaims < StandardError; end
+    
+    attr_reader :payload, :purpose, :expires_at
+    UNIVERSAL_PURPOSE = 'universal'
+
+    def initialize(payload:, **options)
+      @payload = payload
+      @purpose = self.class.pick_purpose(options)
+      @expires_at = pick_expiration(options)
+    end
+
+    class << self
+      attr_accessor :expires_in
+
+      def verify!(claims, options = {})
```

Comment:
I've been thinking about this some more. How about doing it this way:

```ruby
def verify!(claims, options = {})
  claims[:pld] if validate(claims, options, raises: true)
end

def verify(claims, options = {})
  claims[:pld] if validate(claims, options)
end

private
  def validate(claims, options, raises: false)
    if raises
      same_purpose?(claims[:for], options[:for]) || raise InvalidClaims
      fresh?(claims[:exp]) || raise ExpiredClaims
      true
    else
      equal_purpose?(claims[:for], options[:for]) && fresh?(claims[:exp])
    end
  end

  def same_purpose?(claims_purpose, other_purpose)
    claims_purpose == other_purpose || UNIVERSAL_PURPOSE
  end

  def fresh?(expiration)
    Time.iso8601(expiration) < Time.now.utc if expiration
  end
```

Then all the really nasty bits are in `validate`, which can be reused in `verify!` and `verify`. Later if we want to add extra metadata only `validate` has to change.

I'd certainly love it if `validate` could be more readable :smile:

What do you think?